neural_lam/datastore/base.py

```diff
@@ -0,0 +1,479 @@
+        been stacked into a single feature dimension named by the `category` of
+        data being loaded.
+
+        For categories of data that have a time dimension (i.e. not static
+        data), the dataarray is expected additionally have `(analysis_time,
+        elapsed_forecast_duration)` dimensions if `is_forecast` is True, or
+        `(time)` if `is_forecast` is False.
+
+        If the data is ensemble data, the dataarray is expected to have an
+        additional `ensemble_member` dimension.
+
+        Parameters
+        ----------
+        category : str
+            The category of the dataset (state/forcing/static).
+        split : str
+            The time split to filter the dataset (train/val/test).
+
+        Returns
+        -------
+        xr.DataArray or None
+            The xarray DataArray object with processed dataset.
+
+        """
+        pass
+
+    @cached_property
+    @abc.abstractmethod
+    def boundary_mask(self) -> xr.DataArray:
+        """
+        Return the boundary mask for the dataset, with spatial dimensions
+        stacked. Where the value is 1, the grid point is a boundary point, and
+        where the value is 0, the grid point is not a boundary point.
+
+        Returns
+        -------
+        xr.DataArray
+            The boundary mask for the dataset, with dimensions
+            `('grid_index',)`.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_xy(self, category: str) -> np.ndarray:
+        """
+        Return the x, y coordinates of the dataset as a numpy arrays for a
+        given category of data.
+
+        Parameters
+        ----------
+        category : str
+            The category of the dataset (state/forcing/static).
+
+        Returns
+        -------
+        np.ndarray
+            The x, y coordinates of the dataset with shape `[n_grid_points, 2]`.
+        """
+
+    @property
+    @abc.abstractmethod
+    def coords_projection(self) -> ccrs.Projection:
+        """Return the projection object for the coordinates.
+
+        The projection object is used to plot the coordinates on a map.
+
+        Returns
+        -------
+        cartopy.crs.Projection:
+            The projection object.
+
+        """
+        pass
+
+    @functools.lru_cache
+    def get_xy_extent(self, category: str) -> List[float]:
+        """
+        Return the extent of the x, y coordinates for a given category of data.
+        The extent should be returned as a list of 4 floats with `[xmin, xmax,
+        ymin, ymax]` which can then be used to set the extent of a plot.
+
+        Parameters
+        ----------
+        category : str
+            The category of the dataset (state/forcing/static).
+
+        Returns
+        -------
+        List[float]
+            The extent of the x, y coordinates.
+
+        """
+        xy = self.get_xy(category, stacked=False)
+        extent = [xy[0].min(), xy[0].max(), xy[1].min(), xy[1].max()]
```

Comment:
how about we say that for `BaseRegularGridDatastore.get_xy()` default to `stacked=False` so that behaviour of `BaseDatastore.get_xy()` is the same? I will also add to the docstring of `BaseRegularGridDatastore.get_xy()` that by default it should behave like that so that the two implementations behave the same way with default arguments

Reply:
Would that not be `stacked=True` for them to be consistent, and return shape (n_grid_points, 2)?

Reply:
yes, sorry. That is what I meant to write, yes. So by both implementations return stacked grid-coordinates by default

Reply:
OK, this should be done now in: https://github.com/leifdenby/neural-lam/commit/f6da2b21d6f14f14ca8893c27b8af533c40d2bb4